ledger/participant-integration-api/src/main/scala/platform/apiserver/configuration/LedgerConfigProvisioner.scala

```diff
@@ -17,92 +16,87 @@ import com.daml.platform.configuration.LedgerConfiguration
 import com.daml.telemetry.{DefaultTelemetry, SpanKind, SpanName}
 
 import scala.compat.java8.FutureConverters
-import scala.concurrent.Future
 import scala.concurrent.duration.DurationLong
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success, Try}
 
 /** Writes a default ledger configuration to the ledger, after a configurable delay. The
   * configuration is only written if the ledger does not already have a configuration.
   *
   * Used by the participant to initialize a new ledger.
   */
-private[apiserver] final class LedgerConfigProvisioner private (
-    currentLedgerConfiguration: CurrentLedgerConfiguration,
-    writeService: state.WriteConfigService,
-    timeProvider: TimeProvider,
-    submissionIdGenerator: SubmissionIdGenerator,
-    config: LedgerConfiguration,
-    materializer: Materializer,
-)(implicit loggingContext: LoggingContext)
-    extends AutoCloseable {
+object LedgerConfigProvisioner {
   private val logger = ContextualizedLogger.get(getClass)
 
-  private val closed: AtomicBoolean = new AtomicBoolean(false)
-
-  materializer.scheduleOnce(
-    config.initialConfigurationSubmitDelay.toNanos.nanos,
-    () => {
-      if (currentLedgerConfiguration.latestConfiguration.isEmpty && !closed.get)
-        submitInitialConfig(writeService)
-      ()
-    },
-  )
+  def owner(
+      ledgerConfiguration: LedgerConfiguration,
+      currentLedgerConfiguration: CurrentLedgerConfiguration,
+      writeService: state.WriteConfigService,
+      timeProvider: TimeProvider,
+      submissionIdGenerator: SubmissionIdGenerator,
+      scheduler: Scheduler,
+      executionContext: ExecutionContext,
+  )(implicit
+      loggingContext: LoggingContext
+  ): ResourceOwner[Unit] = {
+    implicit val ec: ExecutionContext = executionContext
+    ResourceOwner
+      .forCancellable(() =>
+        scheduler.scheduleOnce(
+          ledgerConfiguration.initialConfigurationSubmitDelay.toNanos.nanos,
+          new Runnable {
```

Comment:
Can't we use a single abstract function here to omit the anonymous class?

Reply:
Unfortunately not. `scheduler.scheduleOnce` is overloaded to accept either a `Runnable` or an `ActorRef`, so Scala doesn't know how to convert the lambda.